### DISCUSSION TITLE
How to generate AND_CERT 4096-RSA

### DISCUSSION THREAD

**Discussion by a9848840** [airsdk/Adobe-Runtime-Support#1866] [Q&A] [answered]
The current certificate in the app is 1024-RSA

Want to upgrade to 4096-RSA

I use Flashdevelop inside CreateCertificate.bat

```
@echo off
cd..
set PAUSE_ERRORS=1
call bat\SetupSDK.bat
call bat\SetupApplication.bat

:: Generate
echo.
echo Generating a self-signed certificate for Android packaging
call adt -certificate -validityPeriod 25 -cn %AND_CERT_NAME% 4096-RSA "%AND_CERT_FILE%" %AND_CERT_PASS%
if errorlevel 1 goto failed

:succeed
echo.
echo Certificate created: %AND_CERT_FILE% with password "%AND_CERT_PASS%"
echo.
if "%AND_CERT_PASS%" == "****" echo (note: you did not change the default password)
echo.
echo HINTS: 
echo - you only need to generate this certificate once,
echo - wait a minute before using this certificate to package your AIR application.
echo.
goto end

:failed
echo.
echo Certificate creation FAILED.
echo.

:end
pause
```

output
Generating a self-signed certificate for Android packaging
key type not recognized: 4096-RSA -- valid type is 2048-RSA

Do I need to modify anything to get 4096-RSA
Or use other tools to generate 4096-RSA

**Comment by ajwfrost** [comment, marked as answer]
Looking at it, currently just 2048-RSA is supported in ADT. We can easily change this to also support 4096-RSA, the issue being that some implementations of the Java runtime may not support this... but I think we can set it up so that if the JRE supports it, then it would work...

But currently -> you may need to use another certificate creation tool.

**Reply by a9848840** [reply]
@ajwfrost 
Do you have any suggested tools?

**Reply by ajwfrost** [reply]
I've used Android Studio before: https://developer.android.com/studio/publish/app-signing#generate-key
but I'm not sure what key length is used there...

There's a tool online: https://certificatetools.com/
I think for that, you'll need to change the "CSR" option to "self-sign", but otherwise it may be good.

Or you can use openssl commands: https://helpcenter.gsx.com/hc/en-us/articles/115015960428-How-to-Generate-a-Self-Signed-Certificate-and-Private-Key-using-OpenSSL

Or, our next release will support certificate generation using RSA-4096 keys...

thanks

**Reply by a9848840** [reply]
Many thanks to ajwfrost for the kind reply

**Reply by a9848840** [reply]
@ajwfrost 
I used opemssl to generate p12
The package file is currently normal
Thank you

openssl req -x509 -newkey rsa:4096 -keyout myKey.pem -out cert.pem -days 36500 -nodes
openssl pkcs12 -export -out ********.p12 -inkey myKey.pem -in cert.pem -passout pass:******
